L1Trigger/TrackFindingTMTT/src/SimpleLR.cc

```diff
@@ -0,0 +1,495 @@
+    }
+  }
+  qOverPt /= invPtToDPhi_;
+
+  bool accepted = false;
+  // double chi2 = (chi2_phi + chi2_z)/2.;
+  // double chi2 = sqrt(chi2_phi*chi2_phi + chi2_z*chi2_z)/2;
+
+  double chi2 = chi2_phi;  // Ignore r-z residuals due to poor 2S resolution?
+  if (digitize_)
+    chi2 = floor(chi2 * chi2Mult_) / chi2Mult_;
+
+  // cout << "chi2 "<< chi2 << " phi "<< chi2_phi << " z "<< chi2_z << endl;
+  float dof = 2 * fitStubs.size() - 4;
+  float chi2dof = chi2 / dof;
+  if (chi2 < chi2cut_)
+    accepted = true;
+
+  if (settings_->debug() == 6)
+    cout << "qOverPt " << qOverPt << " phiT " << phiT << endl;
+
+  // This condition can only happen if cfg param TrackFitCheat = True.
+  if (fitStubs.size() < 4)
+    accepted = false;
+
+  // Kinematic cuts -- NOT YET IN FIRMWARE!!!
+  const float tolerance = 0.1;
+  if (fabs(qOverPt) > 1. / (settings_->houghMinPt() - 0.1))
+    accepted = false;
+  if (fabs(z0) > 20.)
+    accepted = false;
+
+  // Create the L1fittedTrack object
+  const unsigned int hitPattern = 0;  // FIX: Needs setting
+  L1fittedTrack fitTrk(
+      settings_, l1track3D, fitStubs, hitPattern, qOverPt, 0., phi0, z0, tanLambda, chi2_phi, chi2_z, 4, accepted);
+
+  if (settings_->enableDigitize())
+    fitTrk.digitizeTrack("SimpleLR");
+
+  if (settings_->debug() == 6 and digitize_) {
+    cout << "Digitized parameters " << endl;
+    cout << "HT mbin " << int(l1track3D.getCellLocationHT().first) - 16 << " cbin "
+         << int(l1track3D.getCellLocationHT().second) - 32 << " iPhi " << l1track3D.iPhiSec() << " iEta "
+         << l1track3D.iEtaReg() << endl;
+    cout << setw(10) << "First Helix parameters: qOverPt = " << fitTrk.qOverPt() << " oneOver2r "
+         << fitTrk.digitaltrack().oneOver2r() << " (" << floor(fitTrk.digitaltrack().oneOver2r() * qOverPtMult_)
+         << "), phi0 = " << fitTrk.digitaltrack().phi0() << " (" << fitTrk.digitaltrack().iDigi_phi0rel()
+         << "), zT = " << zT << " (" << floor(zT * z0Mult_) << "), tanLambda = " << tanLambda << " ("
+         << floor(tanLambda * tanLambdaMult_) << ")" << endl;
+  }
+
+  if (settings_->debug() == 6) {
+    cout << "FitTrack helix parameters " << int(fitTrk.getCellLocationFit().first) - 16 << ", "
+         << int(fitTrk.getCellLocationFit().second) - 32 << " HT parameters "
+         << int(fitTrk.getCellLocationHT().first) - 16 << ", " << int(fitTrk.getCellLocationHT().second) - 32 << endl;
+
+    if (fitTrk.getMatchedTP() != nullptr) {
+      cout << "VERY GOOD! " << chi2dof << endl;
+      cout << "TP qOverPt " << fitTrk.getMatchedTP()->qOverPt() << " phi0 " << fitTrk.getMatchedTP()->phi0() << endl;
+      if (!accepted)
+        cout << "BAD CHI2 " << chi2 << " chi2/ndof " << chi2dof << endl;
+    } else {
+      cout << "FAKE TRACK!!! " << chi2 << " chi2/ndof " << chi2dof << endl;
+      if (l1track3D.getMatchedTP() != nullptr)
+        cout << "was good" << endl;
+    }
+    cout << "layers in track " << fitTrk.getNumLayers() << endl;
```

Comment:
Not thread safe, please use MessageLogger

Reply:
done